### DISCUSSION TITLE
What is the best way to sample uniformly from an purely open or closed interval.

### DISCUSSION THREAD

**Discussion by abhinavprao** [jax-ml/jax#20527] [Q&A] [answered]
`jax.random.uniform` allows two keywords - `minval` (inclusive) and `maxval` (exclusive) to specify the scope of the form [`minval`, `maxval`) for the unifrom distribution.

What is the best way to achieve purely open scope (`minval`, `maxval`) or purely closed scope [`minval`, `maxval`]?  My network completely breaks at the boundaries, so I would like to avoid it.

Do we have to rely on limiting values/precision of the dtype?
```python
xs = jax.random.uniform(key, 1000, minval = minx+TINIEST_FLOAT, maxval=maxx)
```

**Comment by jakevdp** [comment, marked as answer]
You can see how `jax.random.uniform` is implemented here: https://github.com/google/jax/blob/16b3f00e42122d6be07a5bb1f8025fe4e3d1dd47/jax/_src/random.py#L425-L436

`floats` here are floating point numbers with an exponent of zero, lying between 0 (inclusive) and 1 (exclusive): they are then shifted to be between `minval` (inclusive) and `maxval` (exclusive).

Generating uniform random numbers with an exclusive minimum value would not be easy to do directly: it would essentially involve generating all mantissas *except* zero, which is difficult to do in a non-iterative fashion given source randomness coming from random bits.

The easiest way to do this will probably be using a variation of what you suggested at the end of your question; for example with `minval = jnp.nextafter(minx, jnp.inf)`.

**Reply by abhinavprao** [reply]
Thanks! that makes sense
